File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount

let private nucleotides = [|'A'; 'C'; 'G'; 'T'|]

let private onlyNucleotides (strand:string):bool =
    strand.ToCharArray()
    |> Array.forall (fun (character:char) -> Array.contains character nucleotides)

let private emptyCount:Option<Map<char, int>> =
    [|('A', 0); ('C', 0); ('G', 0); ('T', 0)|]
    |> Map.ofArray
    |> Some

let nucleotideCounts (strand:string):Option<Map<char, int>> =
    match onlyNucleotides strand with
    | true ->
        match strand with
        | "" -> emptyCount
        | _ ->
            strand.ToCharArray()
            |> Array.countBy id
            |> Map.ofArray
            |> Some
    | false -> None

File: solutions/scala/nucleotide-count/1/src/main/scala/NucleotideCount.scala
class DNA(strand:String)
{
    private def nucleotides:Array[Char] = Array[Char]('A', 'C', 'G', 'T')

    private def onlyNucleotides (strand:String):Boolean = strand.toCharArray().forall((character:Char) => nucleotides.contains(character))

    private def emptyCount:Right[Unit, Map[Char, Int]] = Right(Map('A' -> 0, 'C' -> 0, 'G' -> 0, 'T' -> 0))

    def nucleotideCounts:Either[Unit, Map[Char, Int]] =
        onlyNucleotides(this.strand) match
        {
            case true =>
                this.strand match
                {
                    case "" => emptyCount
                    case _ =>
                        def chars:Array[Char] = this.strand.toCharArray()
                        Right(Map('A' -> chars.count((n:Char) => n == 'A'), 'C' -> chars.count((n:Char) => n == 'C'), 'G' -> chars.count((n:Char) => n == 'G'), 'T' -> chars.count((n:Char) => n == 'T')))
                }
            case false => Left(())
        }
}